### PR TITLE
governance types inferred by start functions

### PR DESCRIPTION
**Pull request by turadg**
refs: #7178

## Description

Working on PSM upgrade I got tripped up on some governance facet naming and went down a rabbit hole.

This largely addresses #7178 but I don't want to close that until I'm more confident and have types regression tests.

It also removes `getContractGovernor` from the governed creatorFacet, per an earlier discussion

### Security Considerations

n/a
### Scaling Considerations

n/a
### Documentation Considerations

n/a
### Testing Considerations

CI and some manual spot checking that errors aren't gone due to `any`

### CODE REVIEW

File: packages/governance/src/types-ambient.js [hunk 1]
```diff
@@ -550,43 +550,39 @@
...
  *   all the governed parameters
- * @property {(name: string) => Amount} getAmount
- * @property {(name: string) => Brand} getBrand
- * @property {(name: string) => Instance} getInstance
- * @property {(name: string) => Installation} getInstallation
  * @property {(name: string) => Amount} getInvitationAmount
- * @property {(name: string) => bigint} getNat
- * @property {(name: string) => Ratio} getRatio
- * @property {(name: string) => string} getString
- * @property {(name: string) => any} getUnknown
```
Comment (dckc) [inline packages/governance/src/types-ambient.js:592]: These were a fiction? Or they come back some other way? still reading...

File: packages/governance/src/contractGovernor.js [hunk 2]
```diff
@@ -304,6 +309,8 @@ const start = async (zcf, privateArgs) => {
     });
   };
 
+  /** @type {GovernorCreatorFacet<SF>} */
+  // @ts-expect-error cast
```
Comment (dckc) [inline packages/governance/src/contractGovernor.js:309]: this error is a little surprising; aren't we supplying exactly what the type requires?

File: packages/governance/src/types-ambient.js [hunk 3]
```diff
@@ -590,7 +590,7 @@
  * @property {() => ParamManagerRetriever} getParamMgrRetriever - allows accessing
  *   and updating governed parameters. Should only be directly accessible to the
  *   contractGovernor
- * @property {() => CF} getLimitedCreatorFacet - the creator
+ * @property {() => ERef<CF>} getLimitedCreatorFacet - the creator
```
Comment (dckc) [inline packages/governance/src/types-ambient.js:593]: this was a bug that was hidden behind an `any` somewhere?

File: packages/governance/src/types-ambient.js [hunk 4]
```diff
@@ -704,6 +704,19 @@
...
+ * & ((zcf?: any, pa?: any, baggage?: any) => ERef<{creatorFacet: GovernedCreatorFacet<{}>, publicFacet: GovernedPublicFacet<{}>}>)} GovernableStartFn
+ */
  */
 
+/**
+ * @see {StartedInstanceKit}
+ * @template {GovernableStartFn} SF
+ * @typedef GovernanceFacetKit
+ * Akin to StartedInstanceKit but designed for the results of starting governed contracts
+ * Akin to StartedInstanceKit but designed for the results of starting governed contracts. Used in bootstrap space.
```
Comment (dckc) [inline packages/governance/src/types-ambient.js]: Duplicated line?

File: packages/inter-protocol/src/proposals/econ-behaviors.js [hunk 5]
```diff
@@ -36,16 +36,12 @@ const BASIS_POINTS = 10_000n;
...
  * @property {AdminFacet} psmAdminFacet
  */
 
 /**
- * @typedef {object} AuctioneerKit
- * @property {Awaited<ReturnType<import('../auction/auctioneer.js').start>>['creatorFacet']} creatorFacet
- * @property {Awaited<ReturnType<import('../auction/auctioneer.js').start>>['publicFacet']} publicFacet
- * @property {GovernedContractFacetAccess<{},{}>} governorCreatorFacet
- * @property {AdminFacet} adminFacet
+ * @typedef {GovernanceFacetKit<import('../auction/auctioneer.js').start>} AuctioneerKit
```
Comment (dckc) [inline packages/inter-protocol/src/proposals/econ-behaviors.js:44]: :clap:

File: packages/inter-protocol/src/proposals/econ-behaviors.js [hunk 6]
```diff
@@ -63,24 +59,9 @@ const BASIS_POINTS = 10_000n;
...
- *     adminFacet: AdminFacet,
- *     publicFacet: StakeFactoryPublic,
- *   },
- *   vaultFactoryKit: {
- *     publicFacet: VaultFactoryPublicFacet,
- *     creatorFacet: VaultFactoryCreatorFacet,
- *     governorCreatorFacet: GovernedContractFacetAccess<VaultFactoryPublicFacet, VaultFactoryCreatorFacet>,
- *     adminFacet: AdminFacet,
- *   },
+ *   reserveKit: GovernanceFacetKit<import('../reserve/assetReserve.js')['start']>,
```
Comment (dckc) [inline packages/inter-protocol/src/proposals/econ-behaviors.js:62]: ooh... ahhh...

File: packages/inter-protocol/test/vaultFactory/test-vaultLiquidation.js [hunk 7]
```diff
@@ -305,7 +305,7 @@ test('price drop', async t => {
   const {
     vaultFactory: { vaultFactory, vfPublic },
     priceAuthority,
-    reserveKit: { reserveCreatorFacet },
+    reserveKit: { creatorFacet: reserveCreatorFacet },
```
Comment (dckc) [inline packages/inter-protocol/test/vaultFactory/test-vaultLiquidation.js]: was this a latent bug?

File: packages/governance/src/types-ambient.js [hunk 8]
```diff
@@ -707,13 +707,29 @@
...
+ * @property {AdminFacet} adminFacet
+ * @property {GovernorCreatorFacet<InstallationStart<Awaited<I>>>} creatorFacet
+ * @property {GovernorPublic} publicFacet
  */
 
 /**
  * @see {StartedInstanceKit}
  * @template {GovernableStartFn} SF
  * @typedef GovernanceFacetKit
- * Akin to StartedInstanceKit but designed for the results of starting governed contracts
```
Comment (dckc) [inline packages/governance/src/types-ambient.js]: ah. you found it.

**Review by dckc — APPROVED** [review]
next level stuff!

**Review by dckc — COMMENTED** [review]
I guess we better resolve this before I approve:

```
message: 'Cannot deliver "addIssuer" to target; typeof target is "undefined"'
```

File: packages/governance/src/contractGovernor.js (hunk 2 above)
Comment (turadg) [reply packages/governance/src/contractGovernor.js:309]: It surprised me too. Apparently within the function TS will only trust the minimum of the parameter (what SF is said to extend) which is `{}`. I tried doing this without an error but timed out

File: packages/governance/src/types-ambient.js (hunk 3 above)
Comment (turadg) [reply packages/governance/src/types-ambient.js:593]: it's just not exercised. the code calling it never assumed it to be a presence but this is more accurate

File: packages/inter-protocol/test/vaultFactory/test-vaultLiquidation.js (hunk 7 above)
Comment (turadg) [reply packages/inter-protocol/test/vaultFactory/test-vaultLiquidation.js]: this was to change less in the test file when renaming `reserveCreatorFacet` in `reserveKit`

EDIT: oh but that was erroneous https://github.com/Agoric/agoric-sdk/actions/runs/4759091407/jobs/8458009082?pr=7469

**Pull request merged**